Added parallelization to polymer.PersistenceLength

DrDomenicoMarson opened the pull request:
fix #4671

Changes made in this Pull Request:
- Added parallelization to polymer.PersistenceLength

## PR Checklist
 - [x] Issue raised/referenced?
 - [x] Tests updated/added?
 - [x] Documentation updated/added?
 - [x] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

I haven't added new tests, but I checked that this version pass the relevant tests from the testsuite.


## Developers Certificate of Origin
I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).

<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5074.org.readthedocs.build/en/5074/

<!-- readthedocs-preview mdanalysis end -->

DrDomenicoMarson committed:
added multiprocessing backend

DrDomenicoMarson committed:
added dask backend

DrDomenicoMarson committed:
add versionchanged to docstring

DrDomenicoMarson committed:
fixed versionchanged version

DrDomenicoMarson committed:
added parallell PersistenceLength

DrDomenicoMarson committed:
darker compliant

DrDomenicoMarson committed:
moved to proper section, added PR#

orbeckst commented:
@talagayev would you be able to look after this PR? You have plenty of experience with parallelization. You can try and find reviewers or you can review yourself. Ultimately, you're acting like an editor and you try to shepherd the PR to completion.

If you don't have the bandwidth, please say something and un-assign yourself. Thanks!

talagayev commented:
@orbeckst yes I can review and shepherd it and if I have troubles I will probably ask also Yuxuan and Egor for help :)

orbeckst commented:
@talagayev could you provide an initial review so that @DrDomenicoMarson would have some idea what might need to be done?

orbeckst reviewed (changes requested):
Quick comments: You need to add specific tests for the parallel version. Look into https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/analysis/conftest.py and follow the pattern there. You need to define a `client_PersistenceLength` that will be used for the `run(..., **client_PersistenceLength)` method.

Then you need to update [tests_persistencelength.py](https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/analysis/test_persistencelength.py) to use the new client that you defined. See, for instance, https://github.com/MDAnalysis/mdanalysis/blob/d412c9a9a56312c1bd4e33e6dd3afc4cec7783ca/testsuite/MDAnalysisTests/analysis/test_lineardensity.py#L160 as an example.

Also include an explicit test for the "parallelizable" attribute. Check existing tests for details (e.g. in test_rms.py or test_lineardensity.py).

DrDomenicoMarson committed:
updated the tests for persistencelength to use a client in conftest

DrDomenicoMarson committed:
added test for parallelizability

DrDomenicoMarson commented:
Hi, I updated the tests as requested, and added the test for the "parallelizable" attribute. I'm not sure I quite understood 100% of the testing mechanism (as I'm sort of new in such things), but I ran the tests and all passed on my machine.

talagayev commented:
Ah sorry for the late reply, had to focus on work last week, so I didn't have time to focus on this PR. @DrDomenicoMarson from the first look it looks correct and you adjusted the tests as @orbeckst suggested. I will run and check everything also locally from my side to see if everything works

DrDomenicoMarson committed:
black compliant

orbeckst commented:
While you were working on this pr did you look at open PR https://github.com/MDAnalysis/mdanalysis/pull/4871 ? That PR had stalled but if you used anything then we would also add the original PR author.

orbeckst commented on package/CHANGELOG:
```diff
@@ -58,6 +58,8 @@ Enhancements
    (PR #5038)
  * Moved distopia checking function to common import location in
    MDAnalysisTest.util (PR #5038)
+ * Enables parallelization for `analysis.polymer.PersistenceLength`. (PR #5074)
```
Add the issue number

RMeli replied:
```suggestion
 * Enables parallelization for `analysis.polymer.PersistenceLength` (Issue #4671, PR #5074)
```

orbeckst commented on package/MDAnalysis/analysis/polymer.py:
```diff
@@ -350,7 +369,7 @@ def plot(self, ax=None):
         import matplotlib.pyplot as plt
 
         if ax is None:
-            fig, ax = plt.subplots()
```
Why renamed?

RMeli replied:
I guess maybe because `fig` is not used later on? I don't think it's a necessary change, but if this is the case, the placeholder `_` should probably be used instead.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_persistencelength.py:
```diff
@@ -36,24 +36,33 @@
     def u():
         return mda.Universe(Plength)
 
-    @staticmethod
```
Does it not work with staticmethod?

talagayev replied:
checked it now,

the `@staticmethod` is retained then it will lead to errors due to 

`def p(self, u):` having the `self` as input. It works if we remove the `self` and have:

`def p(u):`

RMeli replied:
That error is expected, since a static method does not take `self` as a parameter. Maybe these should be reverted then?

orbeckst replied:
I probably would have written it as a normal method and not a static method. The only reason why I asked is because I am typically looking for minimal diffs and if anything gets changed I want to know if there's a real reason for it or just aesthetics.

Almost all other tests are normal methods, except the next one, which was also changed from a static method to a normal one. Thus, for overall consistency I now support changing this to a normal method and I'll resolve my comment.

DrDomenicoMarson replied:
Sorry for my late reply, yeah, if I understand the testing approach it should work with and without the 'staticmethod'. I changed it as I saw that in the other tests like test_lineardensity.py staticmethod was not used, let's say I did it for consistency.

orbeckst reviewed (changes requested):
Testing looks good, only minor comments.

talagayev commented:
Looks also good from my side, tests run fine locally.

RMeli commented on testsuite/MDAnalysisTests/analysis/test_persistencelength.py:
```diff
@@ -98,8 +103,8 @@ def test_current_axes(self, p_run):
-    def test(self, p, attr):
-        p_run = p.run(step=3)
+    def test(self, p_run, attr):
+        # p_run = p.run(step=3)
```
```suggestion
```

orbeckst commented on package/MDAnalysis/analysis/polymer.py:
```diff
@@ -350,7 +369,7 @@ def plot(self, ax=None):
 
         if ax is None:
-            fig, ax = plt.subplots()
+            _fig, ax = plt.subplots()
```
```suggestion
            _, ax = plt.subplots()
```

orbeckst committed:
Apply suggestions from code review

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

orbeckst reviewed (changes requested):
@DrDomenicoMarson please include your GH handle in the contributor list for 2.10.0 in CHANGELOG.

Everything else looks good to me! Thank you for the contribution!!

DrDomenicoMarson commented:
> #4671

Hi, sorry I didn't check for existing open PR, so I didn't know someone already worked on this!

DrDomenicoMarson committed:
added my GH handle to CHANGELOG

orbeckst committed:
Merge branch 'develop' into develop

orbeckst reviewed (approved):
LGTM, thank you for your contribution @DrDomenicoMarson !

orbeckst commented:
@talagayev please feel free to squash-merge when all status checks pass. 

For the squash-merge, rewrite the log message with the `- fixes #4671` bullet point and summarize everything else that was done. Basically, the message for the single squashed commit should read nicely.

Thank you for shepherding the PR to completion! Nice work everyone.

yuxuanzhuang commented:
Thanks! @DrDomenicoMarson
